spec/instant.html

```diff
@@ -647,5 +625,30 @@ <h1>TemporalInstantToString ( _instant_, _timeZone_, _precision_ )</h1>
         1. Return the string-concatenation of _dateTimeString_ and _timeZoneString_.
       </emu-alg>
     </emu-clause>
+    <emu-clause id="sec-temporal-differencetemporalinstant" aoid="DifferenceTemporalInstant">
+      <h1>
+        DifferenceTemporalInstant (
```

Comment:
why does this one not have/need a direction?

Reply:
It is easier to just swap the first and second instance instead

Reply:
then why isn't that approach used on all of them?

it seems like consistency between these AOs is worth having - either all with direction or none (i like "none", it's more concise)

Reply:
> then why isn't that approach used on all of them?

because it is not easy to swap for the logic in other types.

> consistency
the logic between these types are not consistent. I thin  simplicity be preferred over  consistency?

Reply:
Then i think they should all take a direction.

Consistency is simplicity, and explicit > implicit - seeing one of these operations and having them all work the same seems simpler to me than having to remember that only one of them swaps the order to indicate sign.

Reply:
I simply disagree

Reply:
Perhaps it'd be helpful if I shared a bit of context. For Instant and PlainTime, `until` and `since` are reversible. `a.until(b)` will always return an equivalent Duration as `b.since(a)`. This works because `until` and `since` for these types only accepts hours and smaller units, which are always the same length.

However, months and (because DST) days aren't always the same length. This means that the starting point affects the result of the operation. 

For example: 
```js
Temporal.PlainDate.from('2022-01-31').until('2022-03-30', { largestUnit: 'month' });
// => P1M30D
Temporal.PlainDate.from('2022-03-30').since('2022-01-31', { largestUnit: 'month' });
// => P1M28D
```

To understand why these results are different, here's what the algorithm is doing:

* 2022-01-31 until 2022-03-30
  * Add 1 month to get Feb 28 (per Temporal's `constrain` behavior)
  * Add 30 days to get Mar 30
  * Result: P1M30D
* 2022-03-30 since 2022-01-31
  * Subtract 1 month to get Feb 28
  * Subtract 28 days to get Jan 31
  * Result: P1M28D

With that context in mind, here's why Frank is making these AOs inconsistent: 
* An AO for `Instant` and `PlainTime` that combines `until` and `since` requires only one bit of information: the direction of the operation. 
* An AO for `until` and `since` for the other types requires two bits of information: the direction of the operation *and* which parameter is the starting point.

So there is a valid reason for `until`/`since` AOs for `Instant` and `PlainTime` to be different than for the other types.

That said, I have concerns:

First, I agree with @ljharb that the inconsistency is confusing. Is there some way (e.g. parameter naming, AO naming, etc.) to clarify to the reader *why* these AOs are inconsistent? 

Second, I'm somewhat skeptical about using `1` and `-1` parameters. These algorithms are already very complex, and it's easy to miss or misunderstand the the subtle irreversibility nuances described above. It may not be clear to the reader which of the two bits of information `1` and `-1` represents.

Could we use an enumeration parameter (like we use in other AOs) instead, and convert the enumerated value to 1 or -1 inside the shared AOs where there's less possibility of confusion?  I think something like this might be clearer: 

```
1. Return ? DifferenceTemporalPlainDate('until', _temporalDate_, _other_, _options_).
```

Reply:
Thank for @justingrant 's comment
Currently, my principle of this refactoring is to reduce the complexity of the spec by "reducing the total number of algorith lines" as possible as I can. This is the way (not necessary you agree with but that is a way) to subjective measure complexity. In other words, if I can code it in one line of algorithm I will treat it as "simpler" than if I express in another way to write the algorith in 3 lines. 

I understand you may want to improve "readibility" of the algorithm and make the spec more "readable". I think that is a good goal. But keep in mind that is a DIFFERENT IMPROVEMENT from the current spec. Comparing with the current spect, the 1 and -1 is implicity hidden there. Using 1 and -1 as paramenter, from my point of view, does not reduce the readability of the spec, just not improving the inherited status quo simply because that was not my motivation of this PR. I am not opposing that, just not what I was originally plan to do.

So.... from my point of view, if you can create a solution use the same number of algorithm to express a better readability, I think that help to simplify the spec. (because I personally measure the simplicity by the total number of pages of this spec if printed, or the total line of algoirthm). Otherwise, if we can switch to use the enum option you suggest, but keep in mind that will increase complexity (because that will increase the total number of lines in algorithm). If the champion want to trade readibility against simplicty, I am ok with that but I personally think it is unncessary since AO is NOT public method interface and not observable and eventually somewhere in the algorithm you still have to convert them into 1 and -1 and whatever the confusion you try to solve will still show up in that line, just moving from parameter of private method (AO) to variable inside private method (inside the algorithm of AO). From my point of view, both are private interface and not important place comparing to real PUBLIC methods. 

Do me a favor, just imaging if you want to do that, how many lines that algorithm line will eventually increase and why is it worth of doing for this particular case.

Reply:
> Could we use an enumeration parameter (like we use in other AOs)

Is such approach really better? For example
https://tc39.es/proposal-temporal/#sec-temporal-differenceisodate
```
f. Let mid be ! [AddISODate](https://tc39.es/proposal-temporal/#sec-temporal-addisodate)(y1, m1, d1, years, 0, 0, 0, "constrain").
```

https://tc39.es/proposal-temporal/#sec-temporal-addisodate
```
3.5.9 AddISODate ( year, month, day, years, months, weeks, days, overflow )
```

my understanding is this "an enumeration parameter" you are referring to is now passing in as a string type. I wonder how do you plan to deal with that in the new AO typed declaration under <h1> syntax (I do not know what is the term to referrring to that syntax, sorry). I feel this "use string as enumeration parameter" more trublesome in 262 spec since there are no places in the spec really to ENUMBERATE (list all possible values and definition of each) it. For example, currently, while you use overflow in several AO, there are no where formally define overflow and all possible value. The closest place is https://tc39.es/proposal-temporal/#sec-temporal-totemporaloverflow
but it simply tell a function read in two possible value, without define what that two values mean sematically. and that is for public enumeration.

I remember there are 1 or 2 pure "private" enum in Temporal spec which is not reading from option bug, but again same issue of confusion and readability remain since there are no place to enumerate them in the spec. and I really hate the idea of passing "string" type and the need of "string comparsion" as what is currently use to express them in spec.  If the string came from the user, it is not avoidable, but in this case, the type you try to express is not from the user, I think using string simply increase unnecessary complexity w/o solving readability.

Reply:
Speaking as a former editor and as an implementer of abstract operations in userland (es-abstract), complexity is definitely not correlated with the number of steps/lines (this is echoed by my personal programming experience as well).

More readable is what's simpler; code that is more concise can easily and often be more complex.

---

I agree spec enums would be clearer than -1/+1, but I didn't think it was a big enough concern to call out, given the larger problem I brought up. In general, the spec does try to use spec-internal types for this kind of thing, and not magic sentinel values.

Reply:
To me the -1/+1 is convenient because we use it as a factor, e.g. _sign_ × _duration_.[[Days]]. But on the other hand we have a few steps that behave differently depending on the sign, e.g.

1. If _sign_ < 0, then ...

which seems a bit more iffy to me, because we are presuming some more context than just these -1 and +1 values. Not so iffy that I would request changes to the PR, but it seems like at least Justin feels more strongly than I do about it.

Maybe we should have an operation like FactorFromSign(_sign_)

1. If _sign_ is `~positive~`, return 1.
2. Return -1.

And then we could have both

1. If _sign_ is `~negative~`, then ...

and

1. Let _factor_ be FactorFromSign(_sign_).
2. ... _factor_ × _duration_.[[Days]] ...

Frank, I agree with you that it's not ideal for those string options to be used instead of enumeration values, it's been a low-priority to-do item for me to replace them with enumerations.

However, like Jordan, I don't agree that optimizing for the smallest number of lines or pages is the right optimization to make in strictly all cases.

Reply:
>Maybe we should have an operation like FactorFromSign(sign)
What is sign and how do you define it?

Reply:
How about `~positive~ or ~negative~`?